Added NFC as a new Transport type ("Transport::Type::kNfc").

OlivierGre opened the pull request:
Added NFC as a new Transport type ("Transport::Type::kNfc").

Implementation done for Android and Linux platforms.

Android implementation is done in java and relies on Android's ISODEP interface (https://developer.android.com/reference/android/nfc/tech/IsoDep) Linux implementation relies on PCSCLite library (https://pcsclite.apdu.fr/) available under BSD-like licence.

In both implementation, the NFC transmission is done asynchronously. If the transmission is successful, OnNfcTagResponse() will be called. Otherwise, OnNfcTagError() will be called.

For the compilation of Chiptool in Linux, it is necessary to install the following package: sudo apt-get install libpcsclite-dev

Added CHIP_DEVICE_CONFIG_ENABLE_NFC_COMMISSIONING.

#### Testing
Tested with [TC-OO-2.1].

bzbarsky-apple commented on src/transport/raw/NFC.cpp:
```diff
@@ -0,0 +1,97 @@
+    VerifyOrReturnError(address.GetTransportType() == Type::kNfc, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mState != State::kNotReady, CHIP_ERROR_INCORRECT_STATE);
+
+    chip::DeviceLayer::Internal::NFCCommissioningMgrImpl().SendToNfcTag(std::move(msgBuf));
```
So this assumes that we have only one NFC "connection" at a time and which connection that is is implicitly known by the NFCCommissioningMgrImpl, no matter what `address` is passed in?  As in, there is actually a singleton "NFC address"?

I know that's how BLE is structured, but is this actually what we want?  It's caused a lot of issued for BLE, and we might want to change the way BLE works to stop having those issues...

OlivierGre replied:
Hi Boris,
I have in fact done this implementation on the same model as BLE implementation (when I have started on this, it was this easiest way to introduce NFC transport). 
That being said, having one NFC connection at a time looks fine to me. Android devices can handle only one NFC connection at a time. Linux NFC reader can probably handle several tags simultaneously but I'm not sure that we need this (at least we don't for Test Harness).

I'm not aware of the issues you mention about this structure based on a singleton so I'm open to any advices.

bzbarsky-apple replied:
The issues with BLE arise when you try to start commissioning.  Should that close existing BLE connections?  Does that include the one you just set up as part of your scan to the new thing you are trying to commission?

The result is a bunch of fragile hacks that break a fair amount, as far as I can tell.

bzbarsky-apple replied:
So what happens if you bring the phone near two NFC tags at once?

OlivierGre replied:
@bzbarsky-apple Sorry for the late answer. 
IMO, when doing NFC Commissioning we expect that there is only one NFC tag in the field (which is the NFC tag present on the device that you want to commission). If there are more than one, we will get an unpredictable results and it is quite dependent of the platform used.
- Android is able to communicate with only one tag at a time. The selection of the tag is unpredictable.
- For Linux implementation, the PCSC-lite implementation that I have used is able to handle a single tag. 

In short, if the customer brings more than one NFC Tag in the field, he will get an unpredictable result.

bzbarsky-apple replied:
Sure, I get that.  If your thing can only deal with one NFC tag, then we start talking to only one, and that's the one we talk to, ok.

The question is whether on platforms that _can_ communicate with more than one NFC tag this behavior will break commissioning completely.  We should not be assuming in this code that the platform can only talk to one NFC tag.

OlivierGre replied:
Ok that makes sense indeed.

I don't feel confortable with this change about NFCCommissioningManager singleton objects. Can I get your advices to decide what to change?

Here is my understanding:

As already discussed, the current implementation is very much based on what is done for BLE. 
In NFCCommissioningManager.h, we have 2 singletons:

```
	/**
	 * Returns a reference to the public interface of the NFCCommissioningManager singleton object.
	 *
	 * Internal components should use this to access features of the NFCCommissioningManager object
	 * that are common to all platforms.
	 */
	extern NFCCommissioningManager & NFCCommissioningMgr();

	/**
	 * Returns the platform-specific implementation of the NFCCommissioningManager singleton object.
	 *
	 * chip applications can use this to gain access to features of the NFCCommissioningManager
	 * that are specific to the selected platform.
	 */
	extern NFCCommissioningManagerImpl & NFCCommissioningMgrImpl();
```

They are implemented in NFCCommissioningManagerImpl and we currently have 2 implementations (one for Android and one for Linux).

* For Android implementation, the things mainly happen in the following files:
src\platform\android\NFCCommissioningManagerImpl.cpp
src\platform\android\java\chip\platform\AndroidNfcCommissioningManager.java

When a message should be sent to NFC Transport (on Android platform), NFCCommissioningManagerImpl.cpp forward the request to AndroidNfcCommissioningManager.java.

AndroidNfcCommissioningManager.java has a function `void setNFCTag(Tag androidTag)` that is used by the application to set the tag to use.
So, if Android platform supported the management of several NFC tags simultaneously, the application could display a popup to ask to the user which one to use and then pass the handle of the tag to AndroidNfcCommissioningManager.java.


* For Linux everything happens in src\platform\linux\NFCCommissioningManagerImpl.cpp
In NFCCommissioningManagerImpl::_Init() we establish the contact with the NFC reader thanks to the PCSC-lite library. 
I think that SCardConnect() can work only if there is a single tag in the field.


So now, let's imagine what we would need if we had a platform supporting multiple NFC tags simultaneously (iOS for instance):
We would have a NFCCommissioningManagerImpl for this platform. I don't know Matter's implementation for iOS. Is it like on android with a native cpp file and an application file?
Is it possible to keep the singleton object and to let the application decide which Tag it passes to NFCCommissioningManagerImpl?

Thanks for your advices on this.

bzbarsky-apple replied:
> Is it like on android with a native cpp file and an application file?

Yes.

But fundamentally: we should not have singletons where multiple things might be going on at once.  For a platform that can talk to multiple NFC tags, it may want to commission multiple devices in parallel.

Just to reiterate: the way this is done for BLE is just architecturally bad, and we have been trying to fix it for years now.  We should not be replicating that mistake.

OlivierGre replied:
> Just to reiterate: the way this is done for BLE is just architecturally bad, and we have been trying to fix it for years now. We should not be replicating that mistake.

My appologize. My point was not to say that we can do it like this "because it is done like that for BLE". I'm sorry if my previous message was not clear! I was trying to explain the current architecture and the impact of this singleton object. I have indicated that the architecture is the same as for BLE to help the other reviewers who jumped into the discussion.

> But fundamentally: we should not have singletons where multiple things might be going on at once. For a platform that can talk to multiple NFC tags, it may want to commission multiple devices in parallel.

Ok I understand and I agree with the goal. However, this is a very significant change and I'm not sure that I will be able to specify this all by myself.

I assume that we want to create one NFCCommissioningManagerImpl instance per NFC Tag. I assume that it is the UID that will be used to distinguisg each tag.
The question is: When shall we instantiate the NFCCommissioningManagerImpl instance?
In current architecture, the NFCCommissioningManagerImpl singleton is invoked when the transport layers are initialized which is at chip stack init. Obviously, we cannot keep that since we don't know yet what will be the NFC Tag to use.

In ChipDeviceController, I have a function called pairDeviceViaNfc(). This function could be changed to receive the UID of the tag that we want to use for the NFC commissioning.
We could store the Tag's UID in Transport::PeerAddress(). By this way NFCBase::SendMessage() (in src/transport/raw/NFC.cpp) would receive the UID of the tag we want to send a message to:
`CHIP_ERROR NFCBase::SendMessage(const Transport::PeerAddress & address, System::PacketBufferHandle && msgBuf)`

NFCBase::SendMessage() would iterate over the existing NFCCommissioningManagerImpls to find one with the requested UID. If it finds one, it would use it. If it doesn't, it would instanciate one.

That's my current naive view but that it's a challenge to do such a big change.

OlivierGre replied:
@bzbarsky-apple Hi Boris,
I was thinking to this discussion and wondering if managing multiple NFC tag is not over enginering:
A platform may want to commission multiple devices in parallel but the NFC phase should be atomic for each one (this is mandatory from tag point of view because it loses all the context when it loses the RF field). So a communication with only one device will happen at a time.
The current solution (with a singleton) doesn't prevent this. The user can do the 1st commissioning phase over NFC for multiple devices sequentially. Then he can power all the devices and the second phase (over thread) will take place.
For a platform able to read multiple tags, the application will have to indicate to NFCCommissioningManagerImpl what NFC tag to use.

bzbarsky-apple replied:
I don't know the details about NFC to tell you for sure but here is my take on this:

1. The API surface used by commissioning should not assume any singletons spanning across commissioning instances.
2. If the implementation of something wants to assume a singleton, because that matches the way things work under the hood, and gracefully handle attempts to do things "in parallel" (e.g. by rejecting them), then it can obviously do so.

bzbarsky-apple replied:
And also: I am agnostic to whether you have one NFCCommissioningManagerImpl or multiple, or what NFCCommissioningManagerImpl even represents.  Again, my request is that the APIs commissioning uses not assume that there is only "one thing", but I also don't have time to design the APIs for this.

You can have a singleton that tracks all the things and is given context to identify which one we are talking about. You can have separate objects for each instance of the "connect and talk to the thing" process.  There are various tradeoffs to both of those approaches, of course.

OlivierGre replied:
@bzbarsky-apple Hi Boris,
I have pushed a new version where the SDK uses the Discriminator to find the NFC Tag it should communicate with. So this version is able to handle several NFC Readers and thus possibly several NFC Tag. 
Please note that the PCSCLite library that I'm using is able to manage only one NFC Tag per reader.
But at least this solution is future proof. An implementation on another platform (iOS for instance) can work even if several tags are in the field.

Can you please review this version?
Thank you

OlivierGre commented on src/transport/raw/NFC.cpp:
```diff
@@ -0,0 +1,97 @@
+
+CHIP_ERROR NFCBase::SendMessage(const Transport::PeerAddress & address, System::PacketBufferHandle && msgBuf)
+{
+#ifdef DUMP_BLE_MESSAGES
```
I'm going to remove those lines.

arnaudrichard commented on scripts/build_python.sh:
```diff
@@ -101,6 +103,14 @@ while (($#)); do
+        --enable_nfc_commissioning | -n)
+            enable_nfc_commissioning=$2
+            if [[ "enable_nfc_commissioning" != "true" && "enable_nfc_commissioning" != "false" ]]; then
+                echo "chip_detail_logging should have a true/false value, not '$enable_nfc_commissioning'"
```
```suggestion
                echo "enable_nfc_commissioning should have a true/false value, not '$enable_nfc_commissioning'"
```

arkq commented on src/include/platform/internal/NFCCommissioningManager.h:
```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2018 Nest Labs, Inc.
```
nit: Nest Labs is the copyright holder of that file, was that copied from nest labs project? Also, for the chip project, the copyright date should indicate when the file was introduced to the project, not the date of project creation. For such file it does not matter, but for some patented algorithm then it is clear whether algorithm was open sourced before the patent claim (in which case patenting such algorithm is not possible) or after that (in which case the usage of such code in a retail product might require a patent fee).

OlivierGre commented:
Linux application are as expected failing on:
```
../third_party/connectedhomeip/src/platform/Linux/NFCCommissioningManagerImpl.cpp:33:10: fatal error: winscard.h: No such file or directory
   33 | #include <winscard.h>
```
since it is necessary to install the following package: 
`sudo apt-get install libpcsclite-dev`

How should I proceed for adding this package? Should I update the bootstrap script? If yes, can I do that in the current PR?

andy31415 commented:
> Linux application are as expected failing on:
> 
> ```
> ../third_party/connectedhomeip/src/platform/Linux/NFCCommissioningManagerImpl.cpp:33:10: fatal error: winscard.h: No such file or directory
>    33 | #include <winscard.h>
> ```
> 
> since it is necessary to install the following package: `sudo apt-get install libpcsclite-dev`
> 
> How should I proceed for adding this package? Should I update the bootstrap script? If yes, can I do that in the current PR?

Update the relevant docker files. If we need them everywhere, then it would probably be in https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/base/chip-build/Dockerfile since all our packages derive from there.

OlivierGre commented:
> Update the relevant docker files. If we need them everywhere, then it would probably be in https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/base/chip-build/Dockerfile since all our packages derive from there.

Thank you very much Andrei, I'm going to do that.

OlivierGre commented:
> Update the relevant docker files. If we need them everywhere, then it would probably be in https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/base/chip-build/Dockerfile since all our packages derive from there.

Andrei, if I edit this Dockerfile locally to add libpcsclite-dev, can I test that it works? When is it supposed to install this package?

When I'm cross-compiling to build the linux-arm64 applications, I'm using a docker image. In this docker image, I had (a long time ago) done a call to "apt-get install libpcsclite-dev" and I had saved the image.

Today, I have removed this package (with "apt-get remove  libpcsclite-dev") but I'm surprised because the build from scratch is still succesful. So I cannot try if the change in the dockerfile that you have indicated is really installing the package libpcsclite-dev. Do you understand why the compilation is still  OK?

OlivierGre commented:
@andy31415 
I have done a PR updating the Dockerfile to add the needed package:
https://github.com/project-chip/connectedhomeip/pull/37137

andy31415 commented on src/transport/raw/BUILD.gn:
```diff
@@ -16,6 +16,7 @@ import("//build_overrides/chip.gni")
 import("${chip_root}/src/ble/ble.gni")
 import("${chip_root}/src/inet/inet.gni")
 import("${chip_root}/src/platform/device.gni")
+import("${chip_root}/src/platform/device.gni")
```
this seems like a duplicate import

andy31415 commented on src/platform/device.gni:
```diff
@@ -88,6 +88,13 @@ declare_args() {
 
+  # Enable NFC Commissioning support
+  if (chip_device_platform == "android" || chip_device_platform == "linux") {
+    chip_enable_nfc_commissioning = true
```
you could do 

```
chip_enable_nfc_commissioning  = (chip_device_platform == "android" || chip_device_platform == "linux");
```

however at the same time, why are we doing this? How about false by default and enable it for build test in a build target.

OlivierGre replied:
When I have written this code, I wanted to implement NFC Transport layer for Android (target android-arm64-chip-tool) and for chip-tool for Linux for the Raspberry Pi used to run the TH (I build the target linux-arm64-chip-tool-ipv6only-clang). 
I understand that there are many linux platforms. I did not want to enable the transport on all those linux platforms so I should update these conditions. Would you have a suggestion to enable chip_enable_nfc_commissioning only for the 2 targets indicated above?

OlivierGre replied:
This code has completely changed in the recent version. The variable is now called chip_enable_nfc_based_commissioning. It is now set to false in device.gni.

andy31415 commented on scripts/build/build/targets.py:
```diff
@@ -170,6 +170,7 @@ def BuildHostTarget():
         'no-interactive', interactive_mode=False).OnlyIfRe('-chip-tool')
     target.AppendModifier("ipv6only", enable_ipv4=False)
     target.AppendModifier("no-ble", enable_ble=False)
+    target.AppendModifier("no-nfc", enable_nfc_commissioning=False)
```
enable_nfc_commissioning already defaults to false in the constructor

andy31415 commented on src/platform/Linux/BUILD.gn:
```diff
@@ -76,6 +76,14 @@ static_library("Linux") {
   ]
 
+  cflags = [
+    "-I/usr/include/PCSC",  # Needed for NFCCommissioningManagerImpl
```
this should not be global. This has to be based on PCSC dependency. 

Also paths should probably be based on pkgconfig and not just added

andy31415 replied:
Please use https://github.com/project-chip/connectedhomeip/blob/master/build/config/linux/pkg_config.gni#L108 and only enable it if `chip_enable_nfc_commissioning`

yunhanw-google commented on src/platform/android/java/chip/platform/AndroidChipPlatform.java:
```diff
@@ -38,6 +38,12 @@ public AndroidChipPlatform(
     initChipStack();
   }
 
+  public native void onNfcTagResponse(byte[] response);
```
Could you also add the UI and implementation in android chip-tool?
https://github.com/project-chip/connectedhomeip/tree/master/examples/android/CHIPTool

Could you also add tests in https://github.com/project-chip/connectedhomeip/tree/master/examples/java-matter-controller/java/src/com/matter/controller/commands

OlivierGre replied:
Hi @yunhanw-google 
I have looked at examples/java-matter-controller/java/src/com/matter/controller/commands but I don't understand what I should do.  Would you have some advices?
Thank you

yunhanw-google replied:
never mind, it seems to be hard to test NFC. do you know whether there is any mock e2e test framework for NFC in android side? thanks

OlivierGre replied:
Sorry, I don't know

yunhanw-google commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
```diff
@@ -0,0 +1,481 @@
+  }
+
+  @Override
+  public void sendToNfcTag(byte[] buf) {
```
Could you follow the below pattern for javadoc on these public API?

/**
 * @brief xxxx
 *
 * This function computes xxx
 *
 * @return xxx
 * @param x xxx
 * @param y xxx
 * @todo xxx
 */

OlivierGre replied:
@yunhanw-google I have pushed an update where I have added javadoc header to NfcCommissioningManager.java and clarified which functions should be public in "AndroidNfcCommissioningManager.java". Please have a look if that's what you expected.
Thank you

andy31415 commented on examples/chip-tool/BUILD.gn:
```diff
@@ -158,3 +158,7 @@ executable("chip-tool") {
   deps = [ ":chip-tool" ]
 }
+
+if (current_os == "linux") {
```
https://gn.googlesource.com/gn/+/main/docs/reference.md#var_target_os seems more correct ...

OlivierGre commented:
@andy31415 Hi Andrei
I'm looking at the CI failure for "Build on Linux (fake, gcc_release, clang, simulated)". The error is :

```
FAIL: test_general_dry_runs (__main__.TestBuilder.test_general_dry_runs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/connectedhomeip/connectedhomeip/out/gcc_release/../../scripts/build/test.py", line 120, in test_general_dry_runs
    self.assertCommandOutput(expected, f'--target {target} --dry-run build'.split(' '))
  File "/__w/connectedhomeip/connectedhomeip/out/gcc_release/../../scripts/build/test.py", line 96, in assertCommandOutput
    self.fail(msg)
AssertionError: DIFFERENCE between expected and generated output in testdata/dry_run_android-arm64-chip-tool.txt
Expected file can be found in dry_run_android-arm64-chip-tool.txt.actual
```
Is it possible to retrieve the files testdata/dry_run_android-arm64-chip-tool.txt and dry_run_android-arm64-chip-tool.txt.actual?

Thanks

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -124,6 +126,43 @@ JNI_METHOD(void, initChipStack)(JNIEnv * env, jobject self)
+
+JNI_METHOD(void, onNfcTagResponse)(JNIEnv * env, jobject self, jbyteArray jbArray)
+{
+    ChipLogProgress(Controller, "%p (Android JNI) onNfcTagResponse()", self);
```
please check whether env is nullptr, thanks

OlivierGre replied:
I'm going to change that.

yunhanw-google commented on src/platform/android/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,141 @@
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+    VerifyOrReturn(env != nullptr, ChipLogError(DeviceLayer, "Failed to GetEnvForCurrentThread for NFCCommissioningManager"));
+
+    mNFCCommissioningManagerObject = env->NewGlobalRef(manager);
```
could you use chip::JniGlobalReference to manage the reference for mNFCCommissioningManagerObject, you can find several examples in sdk , thanks

OlivierGre replied:
Done (thanks for the suggestion!)

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -124,6 +126,46 @@ JNI_METHOD(void, initChipStack)(JNIEnv * env, jobject self)
+{
+    ChipLogProgress(Controller, "%p (Android JNI) onNfcTagResponse()", self);
+
+    VerifyOrReturn(env != nullptr);
```
Please use VerifyOrReturn(env != nullptr, ChipLogError....

OlivierGre replied:
I have done it for all the calls to VerifyOrReturn() that I had done.

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -124,6 +126,46 @@ JNI_METHOD(void, initChipStack)(JNIEnv * env, jobject self)
+    jsize length = env->GetArrayLength(jbArray);
+
+    System::PacketBufferHandle buffer =
+        System::PacketBufferHandle::NewWithData(reinterpret_cast<const uint8_t *>(data), static_cast<size_t>(length));
```
please check VerifyOrExit(!buf.IsNull(), err = CHIP_ERROR_NO_MEMORY);

OlivierGre replied:
Done

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -124,6 +126,46 @@ JNI_METHOD(void, initChipStack)(JNIEnv * env, jobject self)
+    ChipLogProgress(Controller, "%p (Android JNI) onNfcTagResponse()", self);
+
+    VerifyOrReturn(env != nullptr);
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
```
please add JniLocalReferenceScope scope(env); for the jni callback function, for example, 
https://github.com/project-chip/connectedhomeip/blob/ff458d012d392bdf4d5cef0bbee56b5b0fe83c74/src/controller/java/AndroidDeviceControllerWrapper.cpp#L808

yunhanw-google commented on src/platform/android/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,141 @@
+    ChipLogProgress(DeviceLayer, "NFCCommissioningManagerImpl::SendToNfcTag()");
+
+    const uint8_t * buffer = msgBuf->Start();
+    uint32_t len           = (uint32_t) msgBuf->DataLength();
```
please use static_case, thanks

OlivierGre replied:
Done

yunhanw-google commented on src/platform/android/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,141 @@
+
+CHIP_ERROR NFCCommissioningManagerImpl::SendToNfcTag(System::PacketBufferHandle && msgBuf)
+{
+    ChipLogProgress(DeviceLayer, "NFCCommissioningManagerImpl::SendToNfcTag()");
```
SendToNfcTag seems to be called from c++ sdk to java/kotlin layer, please have a  add JniLocalReferenceScope scope(env) which can help manage the lifetime of local reference in case localRef leaks

OlivierGre replied:
Done

yunhanw-google commented on src/platform/android/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,141 @@
+{
+    ChipLogProgress(DeviceLayer, "NFCCommissioningManagerImpl::SendToNfcTag()");
+
+    const uint8_t * buffer = msgBuf->Start();
```
chip::DeviceLayer::StackUnlock unlock; ?

yunhanw-google replied:
not needed. thanks

yunhanw-google commented on src/platform/Linux/NFCCommissioningManagerImpl.h:
```diff
@@ -0,0 +1,131 @@
+
+    // Buffer containing the message to send to the tag
+    uint8_t mDataToSend[1280];
+    uint32_t mDataToSendLength;
```
init it as 0?

OlivierGre replied:
Done in the latest version (which will be pushed soon).

yunhanw-google commented on src/platform/Linux/NFCCommissioningManagerImpl.h:
```diff
@@ -0,0 +1,131 @@
+    uint8_t mAPDUTxBuffer[256];
+    // Buffer used to receive the response to an APDU command
+    uint8_t mAPDURxBuffer[256];
+    uint32_t mAPDUResponseLength;
```
init it as 0?

OlivierGre replied:
Done in the latest version (which will be pushed soon).

yunhanw-google commented on src/platform/Linux/NFCCommissioningManagerImpl.h:
```diff
@@ -0,0 +1,131 @@
+
+    // Buffer storing the chained APDU messages received from the tag
+    uint8_t mChainedResponseBuffer[1280];
+    uint32_t mChainedResponseLength;
```
init it as 0?

OlivierGre replied:
Done in the latest version (which will be pushed soon).

OlivierGre committed:
Added NFC as a new Transport type ("Transport::Type::kNfc").

Implementation done for Android and Linux platforms.

Android implementation is done in java and relies on Android's ISODEP interface
(https://developer.android.com/reference/android/nfc/tech/IsoDep)

Linux implementation relies on PCSCLite library (https://pcsclite.apdu.fr/)
 available under BSD-like licence.

In both implementation, the NFC transmission is done asynchronously.

If the transmission is successful, OnNfcTagResponse() will be called.
Otherwise, OnNfcTagError() will be called.

Added CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING

When building for a platform, nfc-commission should be added to enable
NFC-based commissioning.
Example of command to build chip-tool for linux-arm64:
./scripts/build/build_examples.py --target linux-arm64-chip-tool-ipv6only-clang-nfc-commission build

Testing
Tomorrow, I will push a test checking that NFC-based commissioning works
with PR37060 + PR38540.

bzbarsky-apple commented on src/platform/Linux/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,784 @@
+    {
+        CHIP_ERROR res;
+
+        uint32_t totalLength         = (uint32_t) messageSize;
+        uint32_t nbrOfBytesRemaining = (uint32_t) messageSize;
```
Please don't use C-style casts.

OlivierGre replied:
I will push a new version where I have changed all of them.

bzbarsky-apple commented on src/platform/Linux/NFCCommissioningManagerImpl.cpp:
I did not, generally, review this file; I don't know enough about NFC to do that very quickly.  If no one else has reviewed it either, we should figure out what to do to get it a review.

bzbarsky-apple commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
Did not review this part either.

bzbarsky-apple commented on src/platform/android/NFCCommissioningManagerImpl.cpp:
Did not review this part either.

bzbarsky-apple commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
Did not review any of these Android bits.

bzbarsky-apple commented on src/transport/raw/NFC.cpp:
```diff
@@ -0,0 +1,94 @@
+
+#include <lib/support/BytesToHex.h>
+
+// using namespace chip::Nfc;
```
```suggestion
```

bzbarsky-apple commented on src/transport/raw/NFC.h:
```diff
@@ -0,0 +1,105 @@
+    {
+        kNotReady    = 0, /**< State before initialization. */
+        kInitialized = 1, /**< State after class is connected and ready. */
+        kConnected   = 2, /**< Endpoint connected. */
```
Is kConnected ever used?  And if so, why is kInitialized documented as "connected"?

OlivierGre replied:
It is a copy paste of BLE code. In NFC code, I can see indeed that kConnected  is not used.

bzbarsky-apple replied:
OK, but then why do the docs for "kInitialized" say "connected"?

bzbarsky-apple replied:
Tracked in https://github.com/project-chip/connectedhomeip/issues/38617

bzbarsky-apple commented on src/transport/raw/NFC.h:
```diff
@@ -0,0 +1,105 @@
+     * Ownership of msg is taken over and will be freed at some unspecified time
+     * in the future (once connection succeeds/fails).
+     */
+    CHIP_ERROR SendAfterConnect(System::PacketBufferHandle && msg);
```
This does not seem to actually be implemented.

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -67,6 +68,7 @@ class PeerAddress
     PeerAddress(const Inet::IPAddress & addr, Type type) : mIPAddress(addr), mTransportType(type) {}
     PeerAddress(Type type) : mTransportType(type) {}
     PeerAddress(Type type, NodeId remoteId) : mTransportType(type), mRemoteId(remoteId) {}
+    PeerAddress(Type type, uint16_t shortId) : mTransportType(type), mNFCShortId(shortId) {}
```
This is a weird constructor, because it allows you to pass some other type...  Since we have a factory method for this anyway, can we just make this constructor private, have it just take a shortid, and set mTransportType to NFC in the constructor?

OlivierGre replied:
I have changed this constructor to private. FYI, I want to keep the 2 factory methods:
```
    static PeerAddress NFC() { return PeerAddress(Type::kNfc); }
    static PeerAddress NFC(const uint16_t shortId) { return PeerAddress(Type::kNfc, shortId); }
```
The first one is used on Android side where the Discriminator is not used to find the Tag to use.

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -179,6 +183,9 @@ class PeerAddress
             snprintf(buf, bufSize, "BLE");
             break;
+        case Type::kNfc:
+            snprintf(buf, bufSize, "NFC");
```
Would be good to include mNFCShortId here, no?

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -190,6 +197,8 @@ class PeerAddress
     static PeerAddress Uninitialized() { return PeerAddress(Inet::IPAddress::Any, Type::kUndefined); }
 
     static PeerAddress BLE() { return PeerAddress(Type::kBle); }
+    static PeerAddress NFC() { return PeerAddress(Type::kNfc); }
```
Are there use cases for creating an NFC address without shortid?

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -190,6 +197,8 @@ class PeerAddress
 
     static PeerAddress BLE() { return PeerAddress(Type::kBle); }
+    static PeerAddress NFC() { return PeerAddress(Type::kNfc); }
+    static PeerAddress NFC(const uint16_t shortId) { return PeerAddress(Type::kNfc, shortId); }
```
```suggestion
    static constexpr PeerAddress NFC(const uint16_t shortId) { return PeerAddress(Type::kNfc, shortId); }
```

OlivierGre replied:
I'm getting an error if I do this change:

```
2025-04-24 13:29:20 INFO    In file included from ../../examples/chip-tool/third_party/connectedhomeip/src/transport/raw/Base.h:32:
2025-04-24 13:29:20 INFO    ../../examples/chip-tool/third_party/connectedhomeip/src/transport/raw/PeerAddress.h:200:34: error: constexpr function's return type 'PeerAddress' is not a literal type
2025-04-24 13:29:20 INFO      200 |     static constexpr PeerAddress NFC(const uint16_t shortId) { return PeerAddress(Type::kNfc, shortId); }
2025-04-24 13:29:20 INFO          |                                  ^
2025-04-24 13:29:20 INFO    ../../examples/chip-tool/third_party/connectedhomeip/src/transport/raw/PeerAddress.h:64:7: note: 'PeerAddress' is not literal because it is not an aggregate and has no constexpr constructors other than copy or move constructors
2025-04-24 13:29:20 INFO       64 | class PeerAddress
2025-04-24 13:29:20 INFO          |       ^
2025-04-24 13:29:20 INFO    1 error generated.
```

bzbarsky-apple replied:
Well, the constructor for PeerAddress should be constexpr too, right?

Followup for this, please.

bzbarsky-apple replied:
Tracked in https://github.com/project-chip/connectedhomeip/issues/38614

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -248,6 +257,7 @@ class PeerAddress
     uint16_t mPort               = CHIP_PORT; ///< Relevant for UDP data sending.
     Inet::InterfaceId mInterface = Inet::InterfaceId::Null();
     NodeId mRemoteId             = 0;
+    uint16_t mNFCShortId         = 0;
```
I wonder whether it's worth putting mRemoteId and mNFCShortId in a union.  They're never used together, right?  Followup for that is best.

bzbarsky-apple replied:
Tracked in https://github.com/project-chip/connectedhomeip/issues/38615

OlivierGre committed:
Changed code to use C++ casts.

OlivierGre committed:
Changes following code review.

restyled-commits committed:
Restyled by clang-format

OlivierGre committed:
Added a nullptr check

restyled-commits committed:
Restyled by clang-format

andy31415 committed:
Merge branch 'master' into add_nfc_transport_type

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -245,6 +244,8 @@ class PeerAddress
     }
 
 private:
+    PeerAddress(Type type, uint16_t shortId) : mTransportType(type), mNFCShortId(shortId) {}
```
Followup:
```suggestion
    PeerAddress(uint16_t shortId) : mTransportType(Type::kNfc), mNFCShortId(shortId) {}
```

bzbarsky-apple replied:
Tracked in https://github.com/project-chip/connectedhomeip/issues/38616

robszewczyk commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
```diff
@@ -0,0 +1,488 @@
+
+    if (!mIsoDep.isConnected()) {
+      mIsoDep.close();
+      mIsoDep.connect();
```
Should we be concerned about this lack of mapping between the larger session and the notion of the connection to IsoDep?

OlivierGre replied:
https://github.com/project-chip/connectedhomeip/issues/38833

robszewczyk commented on src/platform/Linux/BUILD.gn:
```diff
@@ -27,6 +27,12 @@ if (chip_enable_openthread) {
 
+if (chip_enable_nfc_based_commissioning) {
+  pkg_config("nfc_commissioning_config") {
+    packages = [ "libpcsclite" ]
```
For other reviewers: `libpcsclite` is primarily licensed under BSD 3 clause license except for the unit tests and for an example code, which appears to be licensed under GPL v.3

robszewczyk commented on src/platform/Linux/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,791 @@
+                //   SW1=0x61 SW2=0xXX if the transmission was successful and the recipient has
+                //     transmitted the first part of a chained response (SW2 indicates the size of
+                //     the data in the next block).
+                if (((sw1 == static_cast<uint8_t>(0x90)) && (sw2 == 0x00)) || (sw1 == static_cast<uint8_t>(0x61)))
```
Why is the static cast needed for 0x90? If this was trigering the compiler errors, I would suggest (for parallelism) also casting 0x00. 

Also a great opportunity to pull out the comparison into a function like `IsStatusSuccess(uint8_t,uint8_t) -> bool`

OlivierGre replied:
Will be fixed in https://github.com/project-chip/connectedhomeip/issues/38832.

robszewczyk commented on src/platform/Linux/NFCCommissioningManagerImpl.cpp:
```diff
@@ -0,0 +1,791 @@
+    //
+    // When the tag's response is fully received, OnNfcTagResponse() notification
+    // will be called.
+    void SendChainedAPDUs(uint8_t * pMessage, size_t messageSize)
```
Is this called from main event processing thread of chip-tool?  It appears to block until the entire message has been sent.

OlivierGre replied:
IMO, this function is not called from main event processing thread of chip-tool.

- NFCCommissioningManagerImpl::SendToNfcTag() is called every times a message should be sent over NFC. This function calls PlatformMgr().ScheduleWork() to schedule a worker thread.
- The worker thread calls this static function:
`void NFCCommissioningManagerImpl::SendToNfcTag(intptr_t arg)`
- which, in turn, calls `void NFCCommissioningManagerImpl::SendChainedAPDUs(intptr_t arg)`.
- This function will then call the function that you mention (SendChainedAPDUs() of TagInstance class).

robszewczyk replied:
I still don't think this works correctly.  filed #39388 to track

OlivierGre commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
```diff
@@ -0,0 +1,488 @@
+          mPlatform.onNfcTagError();
+        }
+      }
+    }.start();
```
I'm going to rework this code.

OlivierGre commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
```diff
@@ -0,0 +1,488 @@
+          mPlatform.onNfcTagError();
+        }
+      }
+    }.start();
```
I'm going to rework this code.

OlivierGre replied:
Will be fixed in https://github.com/project-chip/connectedhomeip/issues/38831

OlivierGre replied:
I have looked at the change that you have done (using workerHandler.post()) and that looks good to me. Thanks for this change.

OlivierGre commented on src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java:
```diff
@@ -0,0 +1,488 @@
+    byte sw1 = response[len - 2];
+    byte sw2 = response[len - 1];
+
+    if ((sw1 == ((byte) 0x90)) && (sw2 == 0x00)) {
```
I'm going to do that.

OlivierGre replied:
Will be fixed in https://github.com/project-chip/connectedhomeip/issues/38832.

robszewczyk replied:
verified this is resolved.

OlivierGre committed:
Fix for Issue #38608:
AndroidNfcCommissioningManager: Use a single worker Thread and use a
message queue.

Fix for Issue #38608.

Change-Id: I15997844fb64292de5533b62d9da22d79eade9bd

OlivierGre committed:
Fix  for issue 38616:
PeerAddess: Change needed to ensure a constructor is only used with
NFC type.

OlivierGre committed:
Fix for Issue #38615 PeerAddress:
Use an Union for mNFCShortId and mRemoteId.

OlivierGre committed:
Fix for Issue #38615

OlivierGre committed:
Fix for Issue #38609 Abstraction of status codes
Fix for Issue #38612 Rework status code management

OlivierGre committed:
Fix for Issue #38617 NFC Transport: Clarify kConnected use

OlivierGre committed:
Fix for Issue #38614 Use constexpr in PeerAddress

Change-Id: I9ac6aadd752d718f01bbc3dcaedac738766d1875

OlivierGre commented:
I have pushed fixes for the following issues:
Issue #38608 AndroidNfcCommissioningManager: Use of a new Thread for each message
Issue #38616 PeerAddess: Change needed to ensure a constructor is only used with NFC type
Issue #38615 PeerAddress: Use an Union for mNFCShortId and mRemoteId
Issue #38609 AndroidNfcCommissioningManager: Abstraction of status codes
Issue #38612 NFCCommissioningManagerImpl: Rework status code management
Issue #38617 NFC Transport: Clarify kConnected use
Issue #38614 Use constexpr in PeerAddress

I have added comments to
Issue #38613 NFCCommissioningManagerImpl: Check execution context
Issue #38611 Check libpcsclite license
Issue #38610 AndroidNfcCommissioningManager: Check IsoDep connection usage

OlivierGre committed:
Merge remote-tracking branch 'csa/master' into add_nfc_transport_type

restyled-commits committed:
Restyled by google-java-format

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on src/transport/raw/NFC.cpp:
```diff
@@ -0,0 +1,110 @@
+        return false;
+    }
+
+    bool canSendToPeer = chip::DeviceLayer::Internal::NFCCommissioningMgrImpl().CanSendToPeer(address);
```
```suggestion
    bool canSendToPeer = DeviceLayer::Internal::NFCCommissioningMgrImpl().CanSendToPeer(address);
```

bzbarsky-apple commented on src/transport/raw/NFC.cpp:
```diff
@@ -0,0 +1,110 @@
+    VerifyOrReturnError(address.GetTransportType() == Type::kNfc, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mState == State::kConnected, CHIP_ERROR_INCORRECT_STATE);
+
+    chip::DeviceLayer::Internal::NFCCommissioningMgrImpl().SendToNfcTag(address, std::move(msgBuf));
```
```suggestion
    DeviceLayer::Internal::NFCCommissioningMgrImpl().SendToNfcTag(address, std::move(msgBuf));
```

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -54,19 +54,25 @@ enum class Type : uint8_t
+    kLast = kNfc, // This is not an actual transport type, it just refers to the last transport type
 };
 
+static constexpr uint16_t kUndefinedNFCShortId = 0;
```
This seems like an implementation detail, no?  Why is this public?  Should be a private member of PeerAddress.

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -54,19 +54,25 @@ enum class Type : uint8_t
-    PeerAddress(const Inet::IPAddress & addr, Type type) : mIPAddress(addr), mTransportType(type) {}
-    PeerAddress(Type type) : mTransportType(type) {}
-    PeerAddress(Type type, NodeId remoteId) : mTransportType(type), mRemoteId(remoteId) {}
+    PeerAddress() : mIPAddress(Inet::IPAddress::Any), mTransportType(Type::kUndefined) { mId.mRemoteId = kUndefinedNodeId; }
```
Followup: we should just make all these constructors constexpr.  But in a separate PR, please.

OlivierGre replied:
PR 39323

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -190,6 +201,8 @@ class PeerAddress
 
     static PeerAddress BLE() { return PeerAddress(Type::kBle); }
+    static constexpr PeerAddress NFC() { return PeerAddress(kUndefinedNFCShortId); }
+    static constexpr PeerAddress NFC(const uint16_t shortId) { return PeerAddress(shortId); }
```
Something needs to document that 0 is not an allowed value for shortId?  At least I assume it's not an allowed value....

OlivierGre replied:
I have pushed a commit with 0xFFFF (+ the other changes)

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -80,7 +86,9 @@ class PeerAddress
     }
 
-    NodeId GetRemoteId() const { return mRemoteId; }
+    NodeId GetRemoteId() const { return mId.mRemoteId; }
```
Should probably document which address types this is valid to call for.  Followup is fine.

OlivierGre replied:
https://github.com/project-chip/connectedhomeip/issues/39349

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -80,7 +86,9 @@ class PeerAddress
-    NodeId GetRemoteId() const { return mRemoteId; }
+    NodeId GetRemoteId() const { return mId.mRemoteId; }
+
+    uint16_t GetNFCShortId() const { return mId.mNFCShortId; }
```
Again, document which address types this is valid to use with.

bzbarsky-apple commented on src/transport/raw/PeerAddress.h:
```diff
@@ -110,7 +118,7 @@ class PeerAddress
     {
         return (mTransportType == other.mTransportType) && (mIPAddress == other.mIPAddress) && (mPort == other.mPort) &&
-            (mInterface == other.mInterface);
+            (mInterface == other.mInterface) && (mId.mNFCShortId == other.mId.mNFCShortId);
```
Should this last check be done only if our transport type is NFC?  Seems like it.

OlivierGre replied:
You're right. Done.

OlivierGre replied:
BTW, I can see that mRemoteId is not part of the comparison. Is it expected?

bzbarsky-apple replied:
Good question.  Looks like that's only used for PAF?  @crlonxp

OlivierGre committed:
Update src/transport/raw/NFC.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

OlivierGre committed:
Update src/transport/raw/NFC.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

OlivierGre committed:
Updates following Boris's review.

OlivierGre committed:
Merge branch 'master' into add_nfc_transport_type

restyled-commits committed:
Restyled by clang-format

OlivierGre committed:
Merge branch 'master' into add_nfc_transport_type

OlivierGre commented on src/transport/raw/PeerAddress.h:
```diff
@@ -117,8 +116,21 @@ class PeerAddress
+            return false;
+        }
+
+        // Compare transport-type-specific fields
```
https://github.com/project-chip/connectedhomeip/issues/39350

woody-apple committed:
Update src/platform/android/java/chip/platform/AndroidNfcCommissioningManager.java

Co-authored-by: Robert Szewczyk <szewczyk@google.com>

restyled-commits committed:
Restyled by google-java-format
